realtek: upgrade to kernel 6.12

plappermaul opened the pull request:
Welcome to cutting edge ...

v8:

- Polished EEE version
- bus hardening
- some splits

v7:

- 1st try of EEE fixup for RTL8218B/D and RTL8214FC

v6:

- patches split completely
- drop unneded refactoring for mac link status to keep patch series small

v5:

- fix typo 

v4:

- split more patches

v3:

- fix for counter warning during secondary cpu startup
- fix for missing XGS-1250 link abilities
- split more patches

v2:

- start splitting the patches
- add missing i2c bus/mux patches

v1:

- For now it runs on my LGS310C
- Big patch needs to be split

plappermaul committed:
realtek: 6.12: create files-6.12 from files-6.6

Automatically generated commit.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: restore files-6.6

Automatically generated commit.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: create patches-6.12 from patches-6.6

Automatically generated commit.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: restore patches-6.6

Automatically generated commit.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: create config-6.12 from config-6.6

Automatically generated commit.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: restore config-6.6

Automatically generated commit.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul commented:
@BrainSlayer @jwh @janh @mrnuke @andigandi @lorand-horvath @svanheule @howels @namiltd @GoetzGoerisch @bmork @Borromini @hardfalcon @evs93 @dangowrt @robimarko just in case you want to join the party again.

plappermaul commented:
This should be self contained and depends on todays realignment patches. I only included the last two outstanding pre-6.12 patches

plappermaul committed:
realtek: 6.12: drop upstreamed otto timer

Thanks to Chris Packham this driver got upstreamed. Drop the
downstream files and patches.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: refresh patches with only source line changes

All these patches needed no manual intervention and applied cleanly
with new source code positions.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: refresh patches with slightly changed anchors

Patches where fuzz had to be cleaned up.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: replace ethtool_eee with ethtool_keee

EEE functions are now called with ethtool_keee instead of
ethtool_eee. Replace all occurrences. This will fix function
signature checks but still produces compilation errors due
to structure changes.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: rename dsa functions/structures to conduit/user

The old upstream notation has been changed to something not so racist.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: add mac capabilities to rtl83xx_phylink_get_caps()

Not only the link but also the mac capabilities are needed here.
Additionally do some alphabetical sorting.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: drop validate from phylink_mac_ops

The validate function no longer exists in phylink_mac_ops. Remove
it for the internal ethernet interface. Instead provide some
meaningful mac capabilities.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: adapt platform patch

Upstream has integrated the Realtek target into the generic MIPS
initialization and so MACH_REALTEK_RTL has gained some new features.
Especially:

- CONFIG_MACH_GENERIC_CORE generates central modules
- board-realtek module adds device specific extensions

The current downstream initialization works well and upgrading to
kernel 6.12 is not the right time to harmonize this. Modify the
MACH definitions to the current needs.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

jwh commented:
Looks good on 1920-48G-POE, not had chance to test others yet, will try to find time tomorrow

howels commented:
Also tested Netgear GS308Tv1 (rtl838x) and HPE JG921A (rtl838x) and both are working including PoE on the JG921A.

So far these rtl83xx targets are working fine but I need to let them run a while and check out VLAN filtering too.

The rtl930x powered Xikestor SKS8300-8X works and appears to be functioning fine at first pass - no obvious regressions with a mix of 1G, 2.5G and 10G SFPs.

Neustradamus commented:
@plappermaul: Good job!

plappermaul committed:
realtek: 6.12: adapt RTL9300 i2c bus & mux drivers

Fix minor compilation errors due to kernel changes.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

vivivitus commented:
Thanks for all your work. I compiled and tested the OpenWRT from your pullrequest today. Basically everything works as before with the exception of all the GbE ports as @jonasjelonek mentioned. I have attached the kernel log below. The SPF+ port is still working and the three 10GbE ports with the RTL8261BE chips are still not working (this is another problem probably due to missing drivers).

<details>
  <summary>Boot Log - XGS1250-12 V2</summary>

  ```
[    0.000000] Linux version 6.12.30 (vivian@openwrt-build-env) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 14.3.0 r29838-ccded0383e) 14.3.0, GNU ld (GNU Binutils) 2.42) #0 Tue May 27 19:37:57 2025
[    0.000000] RTL838X model is 0
[    0.000000] RTL839X model is 0
[    0.000000] RTL93XX model is 93021001
[    0.000000] SoC Type: RTL9302B
[    0.000000] printk: legacy bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019555 (MIPS 34Kc)
[    0.000000] MIPS: machine is Zyxel XGS1250-12 Switch
[    0.000000] earlycon: ns16550a0 at MMIO 0x18002000 (options '115200n8')
[    0.000000] printk: legacy bootconsole [ns16550a0] enabled
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Using appended Device Tree.
[    0.000000] OF: reserved mem: Reserved memory: No reserved-memory node in the DT
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000]   HighMem  empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Kernel command line: earlycon
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Writing ErrCtl register=0000a098
[    0.000000] Readback ErrCtl register=0000a098
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32768
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] RCU Tasks Trace: Setting shift to 0 and lim to 1 rcu_task_cb_adjust=1 rcu_task_cpu_ids=1.
[    0.000000] NR_IRQS: 256
[    0.000000] Failed to get CPU clock: -2
[    0.000000] CPU frequency from device tree: 800MHz
[    0.000000] clocksource: realtek_otto_timer: mask: 0xfffffff max_cycles: 0xfffffff, max_idle_ns: 38225208801 ns
[    0.000002] sched_clock: 28 bits at 3125kHz, resolution 320ns, wraps every 42949672800ns
[    0.009021] Calibrating delay loop... 531.66 BogoMIPS (lpj=2658304)
[    0.065702] pid_max: default: 32768 minimum: 301
[    0.080543] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.088542] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.119596] Memory: 118388K/131072K available (7024K kernel code, 692K rwdata, 1628K rodata, 1208K init, 252K bss, 12164K reserved, 0K cma-reserved, 0K highmem)
[    0.149055] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.159861] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.174285] pinctrl core: initialized pinctrl subsystem
[    0.183126] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.190314] thermal_sys: Registered thermal governor 'step_wise'
[    0.205544] clocksource: Switched to clocksource realtek_otto_timer
[    0.230328] NET: Registered PF_INET protocol family
[    0.236050] IP idents hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.244822] tcp_listen_portaddr_hash hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.254137] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.263671] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.272120] TCP bind hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.279872] TCP: Hash tables configured (established 1024 bind 1024)
[    0.287595] MPTCP token hash table entries: 512 (order: 0, 6144 bytes, linear)
[    0.295877] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.303037] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.311512] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.322684] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.331393] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.337847] jffs2: version 2.2 (NAND) (SUMMARY) (ZLIB) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.354332] pinctrl-single 1b00c600.pinmux: 32 pins, size 4
[    0.361005] pinctrl-single 1b00cc00.pinmux: 32 pins, size 4
[    0.369844] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.378553] printk: legacy console [ttyS0] disabled
[    0.384563] 18002000.uart: ttyS0 at MMIO 0x18002000 (irq = 30, base_baud = 10937500) is a 16550A
[    0.394275] printk: legacy console [ttyS0] enabled
[    0.404981] printk: legacy bootconsole [early0] disabled
[    0.429891] printk: legacy bootconsole [ns16550a0] disabled
[    0.476613] brd: module loaded
[    0.483749] 7 fixed-partitions partitions found on MTD device spi0.0
[    0.490952] OF: Bad cell count for /soc/spi@1200/flash@0/partitions
[    0.498009] OF: Bad cell count for /soc/spi@1200/flash@0/partitions
[    0.505313] Creating 7 MTD partitions on "spi0.0":
[    0.510723] 0x000000000000-0x0000000e0000 : "u-boot"
[    0.518740] 0x0000000e0000-0x0000000f0000 : "u-boot-env"
[    0.526526] 0x0000000f0000-0x000000100000 : "u-boot-env2"
[    0.534356] 0x000000100000-0x000000200000 : "jffs"
[    0.541685] 0x000000200000-0x000000300000 : "jffs2"
[    0.548973] 0x000000300000-0x000000fe0000 : "firmware"
[    0.556769] 2 uimage-fw partitions found on MTD device firmware
[    0.563384] Creating 2 MTD partitions on "firmware":
[    0.568983] 0x000000000000-0x0000003d0000 : "kernel"
[    0.576420] 0x0000003d0000-0x000000ce0000 : "rootfs"
[    0.583798] mtd: setting mtd7 (rootfs) as root device
[    0.589571] 1 squashfs-split partitions found on MTD device rootfs
[    0.596518] 0x000000650000-0x000000ce0000 : "rootfs_data"
[    0.604348] 0x000000fe0000-0x000001000000 : "log"
[    0.624768] Probing RTL838X eth device pdev: 8189a000, dev: 8189a010
[    0.651155] Found SoC ID: 9302: RTL9302B, family 9300
[    0.657016] Using MAC 0000800037d08000
[    0.661249] set sds port 0 to 2
[    0.664773] set sds port 24 to 6
[    0.668419] set sds port 25 to 7
[    0.672016] set sds port 26 to 8
[    0.675657] set sds port 27 to 9
[    0.680048] c45_mask: 000e0000
[    0.791342] REALTEK RTL9300 SERDES mdio-bus:1b: Detected internal RTL9300 Serdes
[    0.808999] i2c_dev: i2c /dev entries driver
[    0.813936] rtl9300_i2c_probe probing I2C adapter
[    0.819281] i2c-rtl9300 1b00036c.i2c-rtl9300: SCL speed 100000, mode is 0
[    0.826885] rtl9300_i2c_probe scl_num 0
[    0.831160] rtl9300_i2c_probe sda_num 1
[    0.839555] NET: Registered PF_INET6 protocol family
[    0.855926] Segment Routing with IPv6
[    0.860175] In-situ OAM (IOAM) with IPv6
[    0.864725] NET: Registered PF_PACKET protocol family
[    0.870912] 8021q: 802.1Q VLAN Support v1.8
[    0.899573] sfp sfp-p12: Host maximum power 1.0W
[    0.976134] REALTEK RTL9300 SERDES rtldsa_mdio-0:1b: Detected internal RTL9300 Serdes
[    0.994523] rtl93xx_setup called
[    0.998248] In rtl83xx_vlan_setup
[    1.001949] UNKNOWN_MC_PMASK: 000000001fffffff
[    1.244922] sfp sfp-p12: module FINISAR CORP.    FCBG110SD1C01    rev R    sn BLCSS11O12473    dc 240704  
[    2.015560] rtl83xx_enable_phy_polling:          f0000ff
[    2.021826] rtl930x_led_init led_set configuration invalid skipping over this set
[    2.030419] rtl83xx-switch switch@1b000000: configuring for fixed/internal link mode
[    2.047661] rtl83xx-switch switch@1b000000 lan1 (uninitialized): validation of xgmii with support 00,00000000,00000000,000062ef and advertisement 00,00000000,00000000,000062c0 failed: -EINVAL
[    2.067902] rtl83xx-switch switch@1b000000 lan1 (uninitialized): failed to connect to PHY: -EINVAL
[    2.077967] rtl83xx-switch switch@1b000000 lan1 (uninitialized): error -22 setting up PHY for tree 0, switch 0, port 0
[    2.098241] rtl83xx-switch switch@1b000000 lan2 (uninitialized): validation of xgmii with support 00,00000000,00000000,000062ef and advertisement 00,00000000,00000000,000062c0 failed: -EINVAL
[    2.118547] rtl83xx-switch switch@1b000000 lan2 (uninitialized): failed to connect to PHY: -EINVAL
[    2.128612] rtl83xx-switch switch@1b000000 lan2 (uninitialized): error -22 setting up PHY for tree 0, switch 0, port 1
[    2.148882] rtl83xx-switch switch@1b000000 lan3 (uninitialized): validation of xgmii with support 00,00000000,00000000,000062ef and advertisement 00,00000000,00000000,000062c0 failed: -EINVAL
[    2.169190] rtl83xx-switch switch@1b000000 lan3 (uninitialized): failed to connect to PHY: -EINVAL
[    2.179259] rtl83xx-switch switch@1b000000 lan3 (uninitialized): error -22 setting up PHY for tree 0, switch 0, port 2
[    2.199603] rtl83xx-switch switch@1b000000 lan4 (uninitialized): validation of xgmii with support 00,00000000,00000000,000062ef and advertisement 00,00000000,00000000,000062c0 failed: -EINVAL
[    2.219837] rtl83xx-switch switch@1b000000 lan4 (uninitialized): failed to connect to PHY: -EINVAL
[    2.229911] rtl83xx-switch switch@1b000000 lan4 (uninitialized): error -22 setting up PHY for tree 0, switch 0, port 3
[    2.250171] rtl83xx-switch switch@1b000000 lan5 (uninitialized): validation of xgmii with support 00,00000000,00000000,000062ef and advertisement 00,00000000,00000000,000062c0 failed: -EINVAL
[    2.270474] rtl83xx-switch switch@1b000000 lan5 (uninitialized): failed to connect to PHY: -EINVAL
[    2.280555] rtl83xx-switch switch@1b000000 lan5 (uninitialized): error -22 setting up PHY for tree 0, switch 0, port 4
[    2.300849] rtl83xx-switch switch@1b000000 lan6 (uninitialized): validation of xgmii with support 00,00000000,00000000,000062ef and advertisement 00,00000000,00000000,000062c0 failed: -EINVAL
[    2.321121] rtl83xx-switch switch@1b000000 lan6 (uninitialized): failed to connect to PHY: -EINVAL
[    2.331181] rtl83xx-switch switch@1b000000 lan6 (uninitialized): error -22 setting up PHY for tree 0, switch 0, port 5
[    2.351558] rtl83xx-switch switch@1b000000 lan7 (uninitialized): validation of xgmii with support 00,00000000,00000000,000062ef and advertisement 00,00000000,00000000,000062c0 failed: -EINVAL
[    2.371765] rtl83xx-switch switch@1b000000 lan7 (uninitialized): failed to connect to PHY: -EINVAL
[    2.381821] rtl83xx-switch switch@1b000000 lan7 (uninitialized): error -22 setting up PHY for tree 0, switch 0, port 6
[    2.402101] rtl83xx-switch switch@1b000000 lan8 (uninitialized): validation of xgmii with support 00,00000000,00000000,000062ef and advertisement 00,00000000,00000000,000062c0 failed: -EINVAL
[    2.422407] rtl83xx-switch switch@1b000000 lan8 (uninitialized): failed to connect to PHY: -EINVAL
[    2.432467] rtl83xx-switch switch@1b000000 lan8 (uninitialized): error -22 setting up PHY for tree 0, switch 0, port 7
[    2.444678] Realtek RTL8261N mdio-bus:18: rtkphy_config_init:69 [RTL8261N/RTL826XB] phy_id: 0x1CCAF3 PHYAD:24
[    3.868792] rtl83xx-switch switch@1b000000 lan9 (uninitialized): PHY [mdio-bus:18] driver [Realtek RTL8261N] (irq=POLL)
[    3.882237] Realtek RTL8261N mdio-bus:19: rtkphy_config_init:69 [RTL8261N/RTL826XB] phy_id: 0x1CCAF3 PHYAD:25
[    5.307055] rtl83xx-switch switch@1b000000 lan10 (uninitialized): PHY [mdio-bus:19] driver [Realtek RTL8261N] (irq=POLL)
[    5.320668] Realtek RTL8261N mdio-bus:1a: rtkphy_config_init:69 [RTL8261N/RTL826XB] phy_id: 0x1CCAF3 PHYAD:26
[    6.741643] rtl83xx-switch switch@1b000000 lan11 (uninitialized): PHY [mdio-bus:1a] driver [Realtek RTL8261N] (irq=POLL)
[    6.755242] rtl83xx-switch switch@1b000000 lan12 (uninitialized): switched to inband/10gbase-r link mode
[    6.767342] rtl838x-eth 1b00a300.ethernet eth0: entered promiscuous mode
[    6.774917] DSA: tree 0 setup
[    6.778365] LINK state irq: 23
[    6.781805] In rtl83xx_setup_qos
[    6.785625] rtl930x_dbgfs_init called
[    6.796305] rtl83xx-switch switch@1b000000: Link is Up - 10Gbps/Full - flow control off
[    6.812168] rtl83xx_fib_event_work_do: FIB4 default rule failed
[    6.819763] clk: Disabling unused clocks
[    6.824709] rtl83xx_fib_event_work_do: FIB4 default rule failed
[    6.837800] VFS: Mounted root (squashfs filesystem) readonly on device 31:7.
[    6.850792] Freeing unused kernel image (initmem) memory: 1208K
[    6.857475] This architecture does not have kernel memory protection.
[    6.864657] Run /sbin/init as init process
[    6.869250]   with arguments:
[    6.872554]     /sbin/init
[    6.875592]   with environment:
[    6.879085]     HOME=/
[    6.881700]     TERM=linux
[    7.385272] init: Console is alive
[    7.389607] init: - watchdog -
[    7.698933] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    7.755266] gpio_button_hotplug: loading out-of-tree module taints kernel.
[    7.766975] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    7.785853] init: - preinit -
[   11.025611] random: crng init done
[   12.214682] RESETTING 9300, CPU_PORT 28
[   12.419330] rtl838x-eth 1b00a300.ethernet eth0: configuring for fixed/internal link mode
[   12.428334] In rtl838x_mac_config, mode 1
[   12.433608] rtl83xx-switch switch@1b000000 lan9: configuring for phy/usxgmii link mode
[   12.442515] rtl93xx_phylink_mac_config SDS is 6
[   12.447935] 8021q: adding VLAN 0 to HW filter on device lan9
[   12.454679] rtl838x-eth 1b00a300.ethernet eth0: Link is Up - 1Gbps/Full - flow control off
[   12.475504] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   12.482931] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   12.497059] rtl83xx_fib_event_work_do: FIB4 failed
[   12.502424] rtl83xx_fib_event_work_do: FIB4 failed
[   12.507865] rtl83xx_fib_event_work_do: FIB4 failed
[   13.775615] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   16.784024] jffs2: notice: (683) jffs2_build_xattr_subsystem: complete building xattr subsystem, 11 of xdatum (0 unchecked, 1 orphan) and 12 of xref (1 dead, 0 orphan) found.
[   16.803315] mount_root: switching to jffs2 overlay
[   16.815940] overlayfs: upper fs does not support tmpfile.
[   16.829567] urandom-seed: Seeding with /etc/urandom.seed
[   16.927449] rtl83xx_fib4_del: no such gateway: 0.0.0.0
[   16.933208] rtl83xx_fib4_del: no such gateway: 0.0.0.0
[   16.946835] rtl83xx_fib4_del: no such gateway: 0.0.0.0
[   16.965658] procd: - early -
[   16.969233] procd: - watchdog -
[   17.653490] procd: - watchdog -
[   17.658309] procd: - ubus -
[   17.722225] procd: - init -
[   18.750578] kmodloader: loading kernel modules from /etc/modules.d/*
[   18.837915] gpio-fan gpio-fan: GPIO fan initialized
[   19.005972] kmodloader: done loading kernel modules from /etc/modules.d/*
[   20.422737] urngd: v1.0.2 started.
[   22.019868] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   29.756129] in rtl838x_eth_stop
[   29.759731] rtl838x-eth 1b00a300.ethernet eth0: Link is Down
[   30.272426] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   30.279813] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   30.287179] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   30.294470] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   30.509528] Using MAC 00007049a208ae4d
[   30.513951] RESETTING 9300, CPU_PORT 28
[   30.718593] rtl838x-eth 1b00a300.ethernet eth0: configuring for fixed/internal link mode
[   30.727596] In rtl838x_mac_config, mode 1
[   30.756285] rtl838x-eth 1b00a300.ethernet eth0: Link is Up - 1Gbps/Full - flow control off
[   30.765694] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   30.773060] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   30.841609] rtl83xx-switch switch@1b000000 lan9: configuring for phy/usxgmii link mode
[   30.850514] rtl93xx_phylink_mac_config SDS is 6
[   30.856174] 8021q: adding VLAN 0 to HW filter on device lan9
[   30.879730] switch: port 1(lan9) entered blocking state
[   30.885648] switch: port 1(lan9) entered disabled state
[   30.891516] rtl83xx-switch switch@1b000000 lan9: entered allmulticast mode
[   30.899251] rtl838x-eth 1b00a300.ethernet eth0: entered allmulticast mode
[   30.907723] rtl83xx-switch switch@1b000000 lan9: entered promiscuous mode
[   31.080815] rtl83xx-switch switch@1b000000 lan10: configuring for phy/usxgmii link mode
[   31.089824] rtl93xx_phylink_mac_config SDS is 7
[   31.095297] 8021q: adding VLAN 0 to HW filter on device lan10
[   31.177042] switch: port 2(lan10) entered blocking state
[   31.182983] switch: port 2(lan10) entered disabled state
[   31.189018] rtl83xx-switch switch@1b000000 lan10: entered allmulticast mode
[   31.251651] rtl83xx-switch switch@1b000000 lan10: entered promiscuous mode
[   31.339463] rtl83xx-switch switch@1b000000 lan11: configuring for phy/usxgmii link mode
[   31.348464] rtl93xx_phylink_mac_config SDS is 8
[   31.353928] 8021q: adding VLAN 0 to HW filter on device lan11
[   31.398713] switch: port 3(lan11) entered blocking state
[   31.404658] switch: port 3(lan11) entered disabled state
[   31.410700] rtl83xx-switch switch@1b000000 lan11: entered allmulticast mode
[   31.446712] rtl83xx-switch switch@1b000000 lan11: entered promiscuous mode
[   31.506029] rtl83xx-switch switch@1b000000 lan12: configuring for inband/10gbase-r link mode
[   31.515455] rtl93xx_phylink_mac_config SDS is 9
[   31.520578] rtl9300_sds_rst 31
[   31.650902] rtl9300_phy_enable_10g_1g 1gbit phy: 00001140
[   31.657959] rtl9300_phy_enable_10g_1g 1gbit phy enabled: 00001140
[   31.665780] rtl9300_phy_enable_10g_1g 10gbit phy: 00002040
[   31.672888] rtl9300_phy_enable_10g_1g 10gbit phy after: 00002040
[   31.680603] rtl9300_phy_enable_10g_1g set medium: 00000002
[   31.687753] rtl9300_phy_enable_10g_1g set medium after: 00000002
[   31.714442] rtl9300_serdes_setup: Configuring RTL9300 SERDES 9
[   31.722981] rtl9300_serdes_mac_link_config: registers before 00000000 00001403
[   31.735059] rtl9300_serdes_mac_link_config: registers after 00000000 00001403
[   31.743043] rtl9300_force_sds_mode: SDS: 9, mode 28
[   31.748519] rtl9300_force_sds_mode --------------------- serdes 9 forcing to 1a ...
[   32.068687] rtl9300_force_sds_mode --------------------- serdes 9 forced to 1a DONE
[   32.082253] start_1.1.1 initial value for sds 9
[   32.144285] end_1.1.1 --
[   32.147145] start_1.1.2 Load DFE init. value
[   32.153889] end_1.1.2
[   32.156454] start_1.1.3 disable LEQ training,enable DFE clock
[   32.174864] end_1.1.3 --
[   32.177737] start_1.1.4 offset cali setting
[   32.184389] end_1.1.4
[   32.186947] start_1.1.5 LEQ and DFE setting
[   32.205634] end_1.1.5
[   32.217191] start_1.2.1 ForegroundOffsetCal_Manual
[   32.226558] end_1.2.1
[   32.237573] start_1.2.3 Foreground Calibration
[   32.256052] rtl9300_do_rx_calibration_2_3: fgcal_gray: 63, fgcal_binary 32
[   32.264714] rtl9300_do_rx_calibration_2_3: end_1.2.3
[   32.270274] start_1.4.1
[   32.493913] end_1.4.1
[   32.496684] start_1.4.2
[   32.509228] vth_set_bin = 2
[   32.511960] vth_set_bin = 2
[   32.516090] Vth Maunal = 0
[   32.635030] Tap0 Sign : +
[   32.638086] tap0_coef_bin = 26
[   32.641998] tap0 manual = 0
[   32.651411] end_1.4.2
[   32.654527] start_1.5.2
[   32.734960] end_1.5.2
[   32.758689] rtl9300_do_rx_calibration: SDS enabled
[   32.819536] start_1.4.1
[   33.043177] end_1.4.1
[   33.045953] start_1.4.2
[   33.058502] vth_set_bin = 2
[   33.061233] vth_set_bin = 2
[   33.065327] Vth Maunal = 0
[   33.184348] Tap0 Sign : +
[   33.187413] tap0_coef_bin = 25
[   33.191332] tap0 manual = 0
[   33.200747] end_1.4.2
[   33.203862] start_1.5.2
[   33.284326] end_1.5.2
[   33.356148] 8021q: adding VLAN 0 to HW filter on device lan12
[   33.395912] rtl83xx-switch switch@1b000000 lan9: Link is Up - 10Gbps/Full - flow control rx/tx
[   33.411749] switch: port 1(lan9) entered blocking state
[   33.417669] switch: port 1(lan9) entered forwarding state
[   33.452779] rtl83xx-switch switch@1b000000 lan12: Link is Up - 10Gbps/Full - flow control off
[   33.689464] switch: port 4(lan12) entered blocking state
[   33.695406] switch: port 4(lan12) entered disabled state
[   33.701449] rtl83xx-switch switch@1b000000 lan12: entered allmulticast mode
[   33.747647] rtl83xx-switch switch@1b000000 lan12: entered promiscuous mode
[   33.755732] switch: port 4(lan12) entered blocking state
[   33.761674] switch: port 4(lan12) entered forwarding state
[   33.842096] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   34.429655] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   34.437072] rtl83xx_fib_event: FIB_RULE ADD/DEL for IPv6 not supported
[   34.462577] rtl83xx-switch switch@1b000000 lan10: Link is Up - 1Gbps/Full - flow control rx/tx
[   34.472319] switch: port 2(lan10) entered blocking state
[   34.478335] switch: port 2(lan10) entered forwarding state
[  112.424047] rtl83xx-switch switch@1b000000 lan10: Link is Down
[  112.430688] switch: port 2(lan10) entered disabled state
  ```
  
</details>

BrainSlayer commented:
> Thanks for all your work. I compiled and tested the OpenWRT from your pullrequest today. Basically everything works as before with the exception of all the GbE ports as @jonasjelonek mentioned. I have attached the kernel log below. The SPF+ port is still working and the three 10GbE ports with the RTL8261BE chips are still not working (this is another problem probably due to missing drivers).
> 
> Boot Log - XGS1250-12 V2

the v2 will work if openwrt merged my changes. i have a v2 and its working. you need a different dts and some code changes in the realtek 10gbit phy driver

all changes are currently just available in the dd-wrt repository but i forwarded the patches already to the openwrt maintainers for integration. technically its easy for you todo it by yourself since you just need todo some small modifications in openwrt

````
[    2.575312] rtl930x_led_init led_set configuration invalid skipping over this set
[    2.575528] rtl83xx-switch switch@1b000000: configuring for fixed/internal link mode
[    2.578260] rtl83xx-switch switch@1b000000 lan01 (uninitialized): PHY [mdio-bus:00] driver [REALTEK RTL8218E] (irq=POLL)
[    2.580387] rtl83xx-switch switch@1b000000 lan02 (uninitialized): PHY [mdio-bus:01] driver [REALTEK RTL8218E] (irq=POLL)
[    2.582608] rtl83xx-switch switch@1b000000 lan03 (uninitialized): PHY [mdio-bus:02] driver [REALTEK RTL8218E] (irq=POLL)
[    2.584841] rtl83xx-switch switch@1b000000 lan04 (uninitialized): PHY [mdio-bus:03] driver [REALTEK RTL8218E] (irq=POLL)
[    2.588640] rtl83xx-switch switch@1b000000 lan05 (uninitialized): PHY [mdio-bus:04] driver [REALTEK RTL8218E] (irq=POLL)
[    2.591527] rtl83xx-switch switch@1b000000 lan06 (uninitialized): PHY [mdio-bus:05] driver [REALTEK RTL8218E] (irq=POLL)
[    2.593757] rtl83xx-switch switch@1b000000 lan07 (uninitialized): PHY [mdio-bus:06] driver [REALTEK RTL8218E] (irq=POLL)
[    2.596707] rtl83xx-switch switch@1b000000 lan08 (uninitialized): PHY [mdio-bus:07] driver [REALTEK RTL8218E] (irq=POLL)
[    2.598935] Realtek RTL8261N mdio-bus:18: rtkphy_config_init:69 [RTL8261N/RTL826XB] phy_id: 0x1CCAF3 PHYAD:24
[    2.599317] rtl83xx-switch switch@1b000000 lan09 (uninitialized): PHY [mdio-bus:18] driver [Realtek RTL8261N] (irq=POLL)
[    2.601176] Realtek RTL8261N mdio-bus:19: rtkphy_config_init:69 [RTL8261N/RTL826XB] phy_id: 0x1CCAF3 PHYAD:25
[    2.601545] rtl83xx-switch switch@1b000000 lan10 (uninitialized): PHY [mdio-bus:19] driver [Realtek RTL8261N] (irq=POLL)
[    2.603430] Realtek RTL8261N mdio-bus:1a: rtkphy_config_init:69 [RTL8261N/RTL826XB] phy_id: 0x1CCAF3 PHYAD:26
[    2.603812] rtl83xx-switch switch@1b000000 lan11 (uninitialized): PHY [mdio-bus:1a] driver [Realtek RTL8261N] (irq=POLL)

````

plappermaul committed:
realtek: 6.12: fix phy-mode for XGS1250-12 port 1-8

Per IEEE 802.3 definition we have:
- parallel XGMII for single 10GBit ONLY links
- serial USGMII for 8 port 1GBit links (not known by kernel)
- serial USXGMII: for single/multiple links with a total bandwidth of 10GBit

The phy-mode of the first eight ports of the XGS1250-12 have always been
defined as XGMII (without S). This came from a confusion with the similar
named Realtek proprietary XSGMII (with S) mode that is basically 10GB SGMII.
From the above definition this is wrong but worked until kernel 6.6. With
the upgrade to 6.12 there is an enforced capabilities check within
phy_caps_from_interface() and link validation fails with

lan1: validation of xgmii with support 62ef and advertisement 62c0 failed: -EINVAL
lan1: failed to connect to PHY: -EINVAL
lan1: error -22 setting up PHY for tree 0, switch 0, port 0

Switch the ports to USXGMII as the most flexible option. This might be no
final solution but at least it better describes the phy/mac link.

Fixes 5b8b382df9a960d88 ("realtek: Add support for ZxXEL XGS1250-12 Switch")
Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: relocate R4K deactivation to late CPU init

To avoid unneeded interrupts the R4K timer is deactivated during
secondary cpu initialization. This is currently done during
phase init_secondary(). With the upgrade to 6.12 the kernel runs
a primary/secondary cpu timer/counter synchronization to verify the
proper setup in synchronise_count_slave(). That runs at a later
point in time and expects the secondary counter to be fully
functional. Finding a deactivated counter results in the following
messages:

WARNING: CPU: 1 PID: 0 at arch/mips/kernel/sync-r4k.c:99 check_counter_warp+0x220/0x254
Warning: zero counter calibration delta: 0 [max: 6500000]
Counter synchronization [CPU#0 -> CPU#1]:
Measured 278760029 cycles counter warp between CPUs

Relocate the deactivation to smp_finsh() at the end of the cpu
startup sequence. Additionally polish the startup code and remove
all unneeded parts.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

jwh commented:
SMP errors have gone, looks to be all good, copper and fibre - thanks!

jonasjelonek commented:
Thanks a lot, XGS1250-12 v1 (and XGS1210-12) now with all ports working again. Will continue testing other stuff on those.

plappermaul committed:
realtek: 6.12: refactor get_mac_eee

Upstream will get rid of the get_mac_eee() function in the DSA
driver and replace it by a boolean alternative. While we fill a
lot of data here (because of EEE bugs in the Realtek phy layer)
other DSA drivers only return if EEE is available or not for a
port. To make the next kernel upgrade easier follow that design.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

hardfalcon commented:
At least in its current form, it seems this patchset doesn't apply cleanly onto head (which is currently commit 48135ddd4457):
```
$ date --utc --iso-8601=seconds ; \
git pull ; \
git show --no-patch --oneline ; \
rm --recursive 'target/linux' ; \
git checkout 'target/linux' ; \
rm --force '18935.patch' ; \
wget 'https://github.com/openwrt/openwrt/pull/18935.patch' ; \
sha256sum '18935.patch' ; \
git apply '18935.patch' ; \
date --utc --iso-8601=seconds
2025-05-29T07:49:49+00:00
Already up to date.
48135ddd4457 (HEAD -> main, origin/master, origin/main, origin/HEAD) generic: 6.12: backport upstream v6.16 b53 patch
Updated 6936 paths from the index
--2025-05-29 09:49:50--  https://github.com/openwrt/openwrt/pull/18935.patch
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://patch-diff.githubusercontent.com/raw/openwrt/openwrt/pull/18935.patch [following]
--2025-05-29 09:49:51--  https://patch-diff.githubusercontent.com/raw/openwrt/openwrt/pull/18935.patch
Resolving patch-diff.githubusercontent.com (patch-diff.githubusercontent.com)... 140.82.121.4
Connecting to patch-diff.githubusercontent.com (patch-diff.githubusercontent.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 200 OK
Cookie coming from patch-diff.githubusercontent.com attempted to set domain to github.com
Length: 1102886 (1,1M) [text/plain]
Saving to: ‘18935.patch’

18935.patch                                          100%[======================================================================================================================>]   1,05M  --.-KB/s    in 0,06s   

2025-05-29 09:49:53 (17,6 MB/s) - ‘18935.patch’ saved [1102886/1102886]

9a6a855e481e136faccf26d0f9fb9c4a2f10985660865781b329bb0270fcdbf3  18935.patch
error: target/linux/realtek/patches-6.6/700-dsa-mdio-increase-max-ports-for-rtl839x-rtl931x.patch: No such file or directory
error: target/linux/realtek/patches-6.12/700-dsa-mdio-increase-max-ports-for-rtl839x-rtl931x.patch: No such file or directory
2025-05-29T07:49:53+00:00
$ 
```

hardfalcon commented:
I've built and booted successfully images for a `TP-Link GS108T v3` and a `TP-Link GS310TP v1` with the changes from this pull request at commit 181f1e9, with (among other changes) the following custom modifications applied on top:
- toolchain: GCC 15.1.0, binutils 2.44, 
- kernel version 6.12.31
- A rebased version of the changes from [linux-hardened 6.12.30-hardened1](https://github.com/anthraxx/linux-hardened/releases/tag/v6.12.30-hardened1)
- mtd-utils 2.3.0

To make the kernel compile with GCC 15, I had to cherry-pick [commit 0f4ae7c6ecb89bfda026d210dcf8216fb67d2333 from `mips-next`](https://git.kernel.org/pub/scm/linux/kernel/git/mips/linux.git/commit/?id=0f4ae7c6ecb89bfda026d210dcf8216fb67d2333).

Another change that I did is adding the following line in `# CONFIG_MIPS_CMDLINE_BUILTIN_EXTEND is not set` in `target/linux/realtek/rtl*x/config-6.12` to allow me to configure a modified kernel cmdline through `CONFIG_CMDLINE` in `target/linux/realtek/rtl*x/config-6.12`. However, this change is nothing specific to your pull request and I've had to add this line with kernel version 5.15 and 6.6 as well.

hardfalcon commented:
I've recompiled the image similar to the description from my last comment, but with this pull request at commit [251b28b](https://github.com/openwrt/openwrt/commit/251b28b73e566d0e80c66282523bb34a96539c47) and with the following pull requests on top:
- #18397 @ 4398d1483ee29798aea8713f4d1a23f5fc1705f7
- #18635 @ f3ad6859746a5635670e02ba55bd1d616a5c73d6
- #18642 @ a83d8d5d688d6e204b205c56b46be7bdc06d3b86
- #18737 @ 9d391636e7f3420b9b36de83af332aa01a434dc6
- #18851 @ 4953c7aa92180a1e10e8efd8d38c0895cc9c8683

Thus far, it appears to work fine on my GS108T v3, however there are a few warnings and errors in `dmesg` (excerpt by using `grep --ignore-case --extended-regexp '(err|fail|warn)'`:
```
[    2.284543] Realtek RTL8218B (internal) mdio-bus:08: ERROR: Port 8 not ready for patch.
[    2.293699] Realtek RTL8218B (internal) mdio-bus:08: probe with driver Realtek RTL8218B (internal) failed with error -1
[    3.895568] Realtek RTL8218B (internal) rtldsa_mdio-0:08: ERROR: Port 8 not ready for patch.
[    3.905212] Realtek RTL8218B (internal) rtldsa_mdio-0:08: probe with driver Realtek RTL8218B (internal) failed with error -1
[    6.863188] rtl83xx_fib_event_work_do: FIB4 default rule failed
[    6.870077] rtl83xx_fib_event_work_do: FIB4 default rule failed
[   29.242799] rtl83xx_fib_event_work_do: FIB4 failed
[   29.248258] rtl83xx_fib_event_work_do: FIB4 failed
[   29.269064] rtl83xx_fib_event_work_do: FIB4 failed
[  144.382718] switch.3: failed to handle neigh update (err -16843010)
[  162.428997] switch.3: failed to handle neigh update (err -16843010)
[  188.188996] switch.3: failed to handle neigh update (err -16843010)
[  330.108897] switch.3: failed to handle neigh update (err -16843010)
[  336.349060] switch.3: failed to handle neigh update (err -16843010)
[  336.356562] switch.3: failed to handle neigh update (err -16843010)
```

I've [attached](https://github.com/user-attachments/files/20507821/dmesg1_redacted.txt) the full `dmesg` output with only part of the MAC address and the build user and build host redacted.

hardfalcon commented on target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/common.c:
```diff
@@ -0,0 +1,1837 @@
+	struct rtl838x_switch_priv *priv;
+	struct net_device *dev;
+	struct neighbour *n = ptr;
+	int err, port;
```
`err` should be initialized to `0`, otherwise it causes spurious `failed to handle neigh update`  messages in `dmesg`.

hardfalcon replied:
I understand your point, and I'm preparing commits/PRs for this in my repo. I would have submitted PRs to fix these issues earlier if I had noticed them earlier.

hailfinger replied:
> I understand your point, and I'm preparing commits/PRs for this in my repo. I would have submitted PRs to fix these issues earlier if I had noticed them earlier.

@hardfalcon is there already a pull request or a tracker bug for this?

plappermaul committed:
realtek: 6.12: align internal/external mac eee function names

The DSA driver uses set_mac_eee() for the outside API while
the interal helper is called port_eee_set(). Align that.
Additionally do not call the internal helpers directly by
the function names but use the register assignments.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: add missing include

The new kernel has relocated the definition of struct platform_device.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: activate testing kernel

Allow to build the new kernel.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

hardfalcon commented on target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -0,0 +1,4069 @@
+		.get_eee	= rtl8218b_get_eee,
+	},
+	{
+		PHY_ID_MATCH_MODEL(PHY_ID_RTL8218B_I),
```
Are you sure the `PHY_ID_RTL8218B_I` in this line is correct? The same value is also used for the preceding `PHY_ID_MATCH_MODEL()` section line 4002.

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -0,0 +1,2685 @@
+	struct resource *res, *mem;
+	phy_interface_t phy_mode;
+	struct phylink *phylink;
+	u8 mac_addr[ETH_ALEN];
```
See above

jwh reviewed (approved):
Looks good on all my test devices

jonasjelonek reviewed (approved):
Looking good and working as expected on my rtl930x-based devices (XGS1250-12, XGS1210-12, TL-ST1008F). Thanks!

howels commented:
Testing passed on my HP JG927A with clean SMP init now.

bmork commented:
runtested on a Netgear GS108Tv3.  Looks good except for eee.  Cannot enable or read state for either end.  No error on config commands, but they have no effect:
```
root@gs108tv3:~# ethtool --show-eee lan3
EEE settings for lan3:
        EEE status: not supported
root@gs108tv3:~# ethtool --set-eee lan3 eee on
root@gs108tv3:~# ethtool --show-eee lan3
EEE settings for lan3:
        EEE status: not supported
```


Funny thing is that eee seems to be working and cannot be disabled.  The other end of this link reports
```
bjorn@rattata:~$ /sbin/ethtool --show-eee eth0
EEE settings for eth0:
        EEE status: enabled - active
        Tx LPI: disabled
        Supported EEE link modes:  100baseT/Full
                                   1000baseT/Full
        Advertised EEE link modes:  100baseT/Full
                                    1000baseT/Full
        Link partner advertised EEE link modes:  100baseT/Full
                                                 1000baseT/Full
```

bmork commented on target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.c:
```diff
@@ -623,35 +623,6 @@ static void rtl838x_port_eee_set(struct rtl838x_switch_priv *priv, int port, boo
-		e->lp_advertised = ADVERTISED_100baseT_Full;
-		e->lp_advertised |= ADVERTISED_1000baseT_Full;
-		return 1;
-	}
```
The removed code reads the RTL838X_MAC_EEE_ABLTY register to detect the **link partner** abilities, not the local mac abilities.  Changing this to report the local mac support is wrong.

Disabling eee on the link partner results in
```
root@gs108tv3:~# ethtool --show-eee lan3
netlink error: Not supported

```

GoetzGoerisch reviewed (approved):
run-tested: HPE 1920-24G (JG924A)

plappermaul committed:
realtek: 6.12: allow mixed c22/c45 bus access

If mmd and normal phy ops are issued the bus is lost because
of wrong park page settings. Force it to 0x1f as in GPL.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: enhance mdio max port patch (once again)

Some bits where missed during the last enhancement of the mdio
patch. In the forthcoming patches the phy_mask will be populated
to avoid unwanted ports (>= cpu port) from being scanned. Add
additional locations where 32 bit values need to be converted
to 64 bits.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: harden the mdio bus even more

With the follow up EEE patches the mdio bus will run c22 and c45
accesses during initial scan. Especially when accessing addresses
beyond the CPU port phy requests might fail in a way that cannot
be handled gratefully. Do two things

- do not allow access to addresses starting from cpu port
- set the scan disable bitmask to ports starting from cpu port

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: proper mdio c45 handling in DSA driver

As the mdio bus has been hardened and can now handle c45 requests
the DSA driver must honor that as well. For this

- add proper upstreamed bus read_c45 and write_c45 functions
- take over the disabled port mask from upstream bus

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: allow fiber media status to be read without lock

rtl8214fc_media_is_fibre() will need to be run when bus lock is held.
Split the function into two versions.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: 6.12: refactor EEE for RTL8218B/D and RTL8214FC

Three different code paths for the same phy model. Now the bus
is prepared to handle c45 (mmd) read/writes correctly. Remove
the custom implementations and let generic kernel functions do
their best. To achieve this

- disable the PHY-mode EEE in rtl821x_config_init() as upstream does
- provide mmd read/write functions that avoid EEE via c45 over c22

While chaning the phy_driver functions sort them alphabetically.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18935
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul commented:
Blocked until https://github.com/openwrt/openwrt/issues/19014 is solved.

plappermaul commented:
https://github.com/openwrt/openwrt/issues/19014 is solved. We can proceed.

GoetzGoerisch reviewed (approved):
run-tested: HPE 1920-24G (JG924A)

- boot from flash (to avoid suprises when everybody only tested initramfs) ☑️
- handling/detection of RTL8214FC ☑️
- EEE stability ☑️

plappermaul commented:
@T-X  feel free to continue your dsa enhancments.

Djfe commented:
awesome, thanks for your continued work @plappermaul

hailfinger commented:
Thank you @plappermaul for lifting the realtek target to 6.12 and thank you @robimarko for merging!

GoetzGoerisch commented:
@plappermaul when should we more to 6.12 as the default kernel for realtek. Any bits/PRs missing?

BrainSlayer commented:
> At least if someone creates such PR 😏

you know my requests. should we do it the real thing or do we want to make dozens other pull requests

BrainSlayer commented:
> > At least if someone creates such PR 😏
> 
> Doing such a PR is not an issue for me. I just want to ask is something is missing before we can/should do this?

we (@plappermaul  and me) are working on bringing all i developed to openwrt. the main progress are rtl931x devices which was the point i stepped in. i have a long backlog of patches which still could be merged but target is first to sync my tree with openwrt for further development. if you want to have fun. review my 6.12 tree. then you see whats different. ethernet driver is faster if you want to use the device as a router and there are specific support for a rtl931x devices. (i just have one for testing).

GoetzGoerisch commented:
> review my 6.12 tree

Do you have a pointer, as it is not on GH?

jonasjelonek commented:
> @GoetzGoerisch Do your duty, be the 6.6 cleaner. Thanks for that.

I was so free and am the 6.6 cleaner in #19139 . Hope noone minds ;)
